src/Catel.MVVM/Collections/FastObservableDictionary.cs

```diff
@@ -0,0 +1,1060 @@
+﻿namespace Catel.Collections
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+    using System.Collections.ObjectModel;
+    using System.Collections.Specialized;
+    using System.ComponentModel;
+    using System.Linq;
+    using System.Runtime.Serialization;
+    using System.Text;
+    using System.Threading.Tasks;
+    using Catel.IoC;
+    using Catel.Logging;
+    using Catel.Services;
+
+    public class FastObservableDictionary<TKey, TValue> : IDictionary<TKey, TValue>,
+        IReadOnlyDictionary<TKey, TValue>,
+        IDictionary,
+        IList<KeyValuePair<TKey, TValue>>,
+        ISerializable,
+        IDeserializationCallback,
+        INotifyCollectionChanged,
+        INotifyPropertyChanged,
+        ISuspendChangeNotificationsCollection
+    {
+        #region Constructors
+        public FastObservableDictionary()
+        {
+            _dict = new Dictionary<TKey, TValue>();
+            _dictIndexMapping = new Dictionary<TKey, int>();
+            _list = new List<TKey>();
+        }
+        public FastObservableDictionary(int capacity)
+        {
+            _dict = new Dictionary<TKey, TValue>(capacity);
+            _dictIndexMapping = new Dictionary<TKey, int>(capacity);
+            _list = new List<TKey>(capacity);
+        }
+        public FastObservableDictionary(IEnumerable<KeyValuePair<TKey, TValue>> originalDict)
+        {
+            if (originalDict is ICollection<KeyValuePair<TKey, TValue>> collection)
+            {
+                _dict = new Dictionary<TKey, TValue>(collection.Count);
+                _dictIndexMapping = new Dictionary<TKey, int>(collection.Count);
+                _list = new List<TKey>(collection.Count);
+            }
+            InsertMultipleValues(0, originalDict, false);
+        }
+        public FastObservableDictionary(IEqualityComparer<TKey> comparer)
+        {
+            _list = new List<TKey>();
+            _dictIndexMapping = new Dictionary<TKey, int>(comparer);
+            _dict = new Dictionary<TKey, TValue>(comparer);
+        }
+        public FastObservableDictionary(IDictionary<TKey, TValue> dictionary)
+        {
+            _dict = new Dictionary<TKey, TValue>(dictionary.Count);
```

Comment:
Can't this call the overload with IEnumerable instead?

Reply:
it's not necessary since the class is un-constructable anyway (the old constructor was internal)

Reply:
they call different overloads of InsertMultipleValues